### PR TITLE
add `geometric` to `random.py`

### PR DESCRIPTION
**Pull request by JiaYaobo**
A counterpart for  `numpy.random.geometric`

### CODE REVIEW

File: jax/_src/random.py [hunk 1]
```diff
@@ -2098,3 +2098,52 @@ def _wald(key, mean, shape, dtype) -> Array:
...
+              shape: Optional[Shape] = None,
+              dtype: DTypeLikeInt = dtypes.int_) -> Array:
+  r"""Sample Geometric random values with given shape and float dtype.
+
+  The values are returned according to the probability mass function:
+
+  .. math::
+      f(k;p) = p(1-p)^{k-1}
+
+  on the domain :math:`0 < p < 1` is the probability of success of an individual trial
```
Comment (jakevdp) [inline jax/_src/random.py:2113]: Delete "is the probability of success of an individual trial" ?

File: jax/_src/random.py [hunk 2]
```diff
@@ -2098,3 +2098,52 @@ def _wald(key, mean, shape, dtype) -> Array:
...
+    shape = core.canonicalize_shape(shape)
+  return _geometric(key, p, shape, dtype)
+
+@partial(jit, static_argnums=(2, 3), inline=True)
+def _geometric(key, p, shape, dtype) -> Array:
+  if shape is None:
+    shape =  np.shape(p)
+  else:
+    _check_shape("geometric", shape, np.shape(p))
+  p = lax.convert_element_type(p, np.float32)
```
Comment (jakevdp) [inline jax/_src/random.py:2144]: Instead of hard-coding float32, we should use
```python
check_arraylike("geometric", p)
p, = promote_dtypes_inexact(p)
```

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks!

**Review by jakevdp — APPROVED** [review]
Clicking "Approve" to trigger additional tests.

File: jax/_src/random.py (hunk 2 above)
Comment (JiaYaobo) [reply jax/_src/random.py]: good advice! Please take a look at poisson related functions in random.py may use the hard-coding too

**Review by jakevdp — APPROVED** [review]

File: jax/_src/random.py [hunk 3]
```diff
@@ -2098,3 +2098,53 @@ def _wald(key, mean, shape, dtype) -> Array:
...
+  return _geometric(key, p, shape, dtype)
+
+@partial(jit, static_argnums=(2, 3), inline=True)
+def _geometric(key, p, shape, dtype) -> Array:
+  if shape is None:
+    shape =  np.shape(p)
+  else:
+    _check_shape("geometric", shape, np.shape(p))
+  check_arraylike("geometric", p)
+  p = promote_dtypes_inexact(p)
```
Comment (jakevdp) [inline jax/_src/random.py:2145]: THis needs to be
```
p, = promote_dtypes_inexact(p)
```

File: jax/_src/random.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/random.py]: Yeah I'm in the process of changing all of those, see https://github.com/google/jax/pull/15177. It's just taking a while to land because it breaks some corner cases.

**Comment by jakevdp** [comment]
Looks like there are some precision issues in the tests

**Review by jakevdp — APPROVED** [review]
Thanks!
